node/cmd/ccq/p2p.go

```diff
@@ -0,0 +1,267 @@
+			if err != nil {
+				logger.Error("received invalid message", zap.Binary("data", envelope.Data),
+					zap.String("from", envelope.GetFrom().String()))
+				continue
+			}
+			switch m := msg.Message.(type) {
+			case *gossipv1.GossipMessage_SignedQueryResponse:
+				logger.Debug("query response received", zap.Any("response", m.SignedQueryResponse))
+				var queryResponse common.QueryResponsePublication
+				err := queryResponse.Unmarshal(m.SignedQueryResponse.QueryResponse)
+				if err != nil {
+					logger.Error("failed to unmarshal response", zap.Error(err))
+					continue
+				}
+				requestSignature := hex.EncodeToString(queryResponse.Request.Signature)
+				// Check that we're handling the request for this response
+				pendingResponse := pendingResponses.Get(requestSignature)
+				if pendingResponse == nil {
+					logger.Debug("skipping query response for unknown request", zap.String("signature", requestSignature))
+					continue
+				}
+				// Make sure that the request bytes match
+				if !bytes.Equal(queryResponse.Request.QueryRequest, pendingResponse.req.QueryRequest) ||
+					!bytes.Equal(queryResponse.Request.Signature, pendingResponse.req.Signature) {
+					continue
+				}
+				digest := common.GetQueryResponseDigestFromBytes(m.SignedQueryResponse.QueryResponse)
+				signerBytes, err := ethCrypto.Ecrecover(digest.Bytes(), m.SignedQueryResponse.Signature)
+				if err != nil {
+					logger.Error("failed to verify signature on response",
+						zap.String("digest", digest.Hex()),
+						zap.String("signature", hex.EncodeToString(m.SignedQueryResponse.Signature)),
+						zap.Error(err))
+					continue
+				}
+				signerAddress := ethCommon.BytesToAddress(ethCrypto.Keccak256(signerBytes[1:])[12:])
+
+				mu.RLock()
+				keyIdx, hasKeyIdx := guardianSet.KeyIndex(signerAddress)
+				q := quorum
+				mu.RUnlock()
+
+				if hasKeyIdx {
+					if _, ok := responses[requestSignature]; !ok {
+						responses[requestSignature] = make(map[ethCommon.Hash][]GuardianSignature)
+					}
+					found := false
+					for _, gs := range responses[requestSignature][digest] {
+						if gs.Index == keyIdx {
+							found = true
+							break
+						}
+					}
+					if found {
+						// Already handled the response from this guardian
+						continue
+					}
+					responses[requestSignature][digest] = append(responses[requestSignature][digest], GuardianSignature{
+						Index:     keyIdx,
+						Signature: hex.EncodeToString(m.SignedQueryResponse.Signature),
+					})
+					// quorum is reached when a super-majority of guardians have signed a response with the same digest
+					if len(responses[requestSignature][digest]) >= q {
+						s := &SignedResponse{
+							Response:   &queryResponse,
+							Signatures: responses[requestSignature][digest],
+						}
+						delete(responses, requestSignature)
+						pendingResponse.ch <- s
+					}
+				} else {
+					logger.Warn("received observation by unknown guardian - is our guardian set outdated?",
```

Comment:
I wonder if this should flip a message to the guardian set routine to make it read immediately? Maybe not worth attempting to retry this message, but might reduce the length of an outage due to a guardian set update.

Reply:
We decided to just fetch the guardian set once at startup. For now the plan is to have some sort of manual intervention during guardian set upgrades